File: typescript/src/two-pointers/valid-palindrome.ts
export function isPalindrome(s: string): boolean {
  let start = 0;
  let end = s.length - 1;

  while (start < end) {
    const a = s[start]!;
    const b = s[end]!;
    if (!isAlphanumeric(a)) {
      start++;
    } else if (!isAlphanumeric(b)) {
      end--;
    } else if (a.toLowerCase() !== b.toLowerCase()) {
      return false;
    } else {
      start++;
      end--;
    }
  }
  return true;
}

function isAlphanumeric(character: string): boolean {
  return (
    ('a' <= character && character <= 'z') ||
    ('A' <= character && character <= 'Z') ||
    ('0' <= character && character <= '9')
  );
}

File: python/src/two_pointers/valid_palindrome.py
def is_palindrome(s: str) -> bool:
    clean = [c for c in s.lower() if c.isalnum()]
    return clean == clean[::-1]


def is_palindrome_two_pointer(s: str) -> bool:
    start = 0
    end = len(s) - 1

    while start < end:
        if not s[start].isalnum():
            start += 1
        elif not s[end].isalnum():
            end -= 1
        elif s[start].lower() != s[end].lower():
            return False
        else:
            start += 1
            end -= 1

    return True


